sui-execution/latest/sui-move-natives/src/crypto/poseidon.rs

```diff
@@ -0,0 +1,103 @@
+    pop_arg,
+    values::{Value, VectorRef},
+};
+use smallvec::smallvec;
+use std::collections::VecDeque;
+use std::ops::Mul;
+
+pub const NON_CANONICAL_INPUT: u64 = 0;
+pub const EMPTY_INPUT: u64 = 1;
+
+#[derive(Clone)]
+pub struct PoseidonBN254CostParams {
+    /// Base cost for invoking the `poseidon_bn254` function
+    pub poseidon_bn254_cost_base: Option<InternalGas>,
+    /// Cost per block of `data`, where a block is 32 bytes
+    pub poseidon_bn254_data_cost_per_block: Option<InternalGas>,
+}
+
+/***************************************************************************************************
+ * native fun poseidon_bn254
+ * Implementation of the Move native function `poseidon::poseidon_bn254(data: &vector<u256>): u256
+ *   gas cost: poseidon_bn254_cost_base                           | base cost for function call and fixed opers
+ *              + poseidon_bn254_data_cost_per_block * num_inputs | cost depends on number of inputs
+ **************************************************************************************************/
+pub fn poseidon_bn254(
+    context: &mut NativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> PartialVMResult<NativeResult> {
+    // Load the cost parameters from the protocol config
+    let cost_params = &context
+        .extensions()
+        .get::<NativesCostTable>()
+        .poseidon_bn254_cost_params
+        .clone();
+
+    // Charge the base cost for this operation
+    native_charge_gas_early_exit!(
+        context,
+        cost_params
+            .poseidon_bn254_cost_base
+            .ok_or_else(
+                || PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Gas cost for poseidon_bn254 not available".to_string())
+            )?
+    );
+
+    debug_assert!(ty_args.is_empty());
+    debug_assert!(args.len() == 1);
+
+    // The input is a reference to a vector of vector<u8>'s
+    let inputs = pop_arg!(args, VectorRef);
+    let length = inputs.len(&Type::U256)?.value_as::<u64>()?;
+
+    if length == 0 {
+        return Ok(NativeResult::err(context.gas_used(), EMPTY_INPUT));
+    }
+
+    // Charge the msg dependent costs
+    native_charge_gas_early_exit!(
+        context,
+        cost_params
+            .poseidon_bn254_data_cost_per_block
+            .ok_or_else(
+                || PartialVMError::new(StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR)
+                    .with_message("Gas cost for poseidon_bn254 not available".to_string())
+            )?
+            .mul(length.into())
+    );
+
+    // Read the input vector and convert each element to a field element in le representation
+    let mut field_elements: Vec<Vec<u8>> = Vec::new();
+    for i in 0..length {
+        let reference: Reference = inputs.borrow_elem(i as usize, &Type::U256)?.cast()?;
+        let value = reference.read_ref()?.value_as::<U256>()?;
+        field_elements.push(value.to_le_bytes().to_vec());
+    }
```

Comment:
For extra fancy points
```suggestion
    let field_elements = (0..length).map(|i| {
        let reference: Reference = inputs.borrow_elem(i as usize, &Type::U256)?.cast()?;
        let value = reference.read_ref()?.value_as::<U256>()?;
        value.to_le_bytes().to_vec();
    ))
    .collect()
```
But at minimum probably best practice to at least switch the `Vec::new()` to `Vec::with_capacity` since the exact size is known